ENH: expose datetime.c functions to cython

jbrockmendel opened the pull request:
xref #9675 would be nice to cimport these rather than duplicating them in pandas.

Need to add a `cdef extern from "numpy/foobar.h":` for these in `__init__.pxd`, need guidance and what that foobar.h should be.

Is there a standard way of testing what is exposed to cython?  cc @bashtage

jbrockmendel committed:
ENH: expose datetime.c functions to cython

mattip commented:
> Is there a standard way of testing what is exposed to cython

We have cython tests in [`numpy/core/tests/test_cython.py`](https://github.com/numpy/numpy/blob/main/numpy/core/tests/test_cython.py) which builds and imports the [`checks.pyx`](https://github.com/numpy/numpy/blob/main/numpy/core/tests/examples/cython/checks.pyx) cython file. You could add to there or add a new file in `numpy/core/tests/examples`.

> what that `foobar.h` should be

Maybe datetime64_c.h in [`numpy/core/include/numpy/`](https://github.com/numpy/numpy/tree/main/numpy/core/include/numpy)? This would be similar to `numpy/core/include/numpy/random/distributions.h`

I still would prefer we declare datetime64 to be a user-defined dtype, and split the code out into a separate PyPI-installable package. This would be a good test of the new dtype machinery as well as clean up NumPy internals.

jbrockmendel commented on numpy/core/src/multiarray/datetime.c:
```diff
@@ -2035,7 +2033,7 @@ metastr_to_unicode(PyArray_DatetimeMetaData *meta, int skip_brackets)
- */
+
 NPY_NO_EXPORT void
 add_seconds_to_datetimestruct(npy_datetimestruct *dts, int seconds)
```
disabled this function entirely in last commit to confirm it is unused.  any objection to ripping it out?  OK to consider this out of scope.

WillAyd replied:
FYI another way that might help check this would be to declare the function as static and use gcc's -Wunused-function flag
That flag might be a part of -Wall or -Wextra so you might not need an extra flag at all, but it only warns for static functions

seberg replied:
Ah, sorry for not noticing this.  I am happy to delete it here or otherwise.  The issue is that we have `NPY_NO_EXPORT`  in a few places too much.  It should be `static` and then our CI would already take note.

jbrockmendel committed:
tests

seberg commented:
The random number approach seems reasonable on first sight.  I am a bit curious how it compares to most of the API which uses the API table rather than relying on linking?

Splitting out `datetime` would be cool and it might also be a good time for some cleanups.  I.e. we could change some things and keep the current datetimes unmodified and deprecated for a while.
To be clear, I would love a new external DType approach and I do expect it is the right way forward and I am willing to do most of the ground-work related to porting the current datetimes.
However, I don't want to stand in the way of more pragmatic solutions, since I don't know the full picture in pandas.  Also a user DType will require at least the next NumPy version to actually function (which may just be OK for you?).

jbrockmendel commented:
> I am a bit curious how it compares to most of the API which uses the API table rather than relying on linking?

This is way outside my region of competence and I'll definitely need help with... whatever it means.

> Splitting out datetime would be cool and it might also be a good time for some cleanups

Depending on how much of the datetime code we're talking about splitting out, at some point we'd need to reconcile with the place/places where pandas intentionally diverges from the numpy behavior.  Most of those discussions occurred before my time so I don't have a good read on what that would entail.

> Also a user DType will require at least the next NumPy version to actually function (which may just be OK for you?).

Even if this PR were fixed and merged today, it would take a year or two before pandas could actually rely on it and rip out our own code (minimum supported numpy on our last release was 1.18.5).  So I'm not at all concerned about having to wait a while.

seberg commented:
>> I am a bit curious how it compares to most of the API which uses the API table rather than relying on linking?

>This is way outside my region of competence and I'll definitely need help with... whatever it means.

That question was for @mattip, since he was directly involved in exposing the random number API :).

> Depending on how much of the datetime code we're talking about splitting out, at some point we'd need to reconcile with the place/places where pandas intentionally diverges from the numpy behavior

I was talking about splitting it out completely.  NumPy would keep its current ones around for backcompat probably, but we would deprecate it soon after, and point to the new DType defined outside of NumPy.

That is the question for you :).  We could start on new DTypes for datetimes/timedelta.  Obviously, it would be targeted to do whatever pandas wants.  But, both times and pandas are complicated, so while I could take care of most of the DType part we will need to work together and create plan.  And I would probably need help with many of the time related things to make good progress.  My guess is something like:
* Figure out the basic requirements (timezones? resolution?).  One way to solve this for me would be if we had "scalars" for blueprints (that could be pandas scalars or even `datetime.datetime`)
* Create a corresponding DType  (this I can do, maybe not full featured, but with the basics there)
* Fixes, fixes, fixes:
  * Pandas won't work with it out of the box, I have no idea how hard that will be
  * There will be holes in the NumPy API that need to get fixed  (I should be able to do most of this of course)  
* "Pandas adoption": Even if pandas works OK with the new dtype, there is probably quite a bit of things that need to work with the new dtype explicitly.  Maybe that is easy (dtype is very similar to the old one anyway) or maybe not.

The problem is that it seems hard to tell how quickly this can materialize into something usable.  I don't think a prototype will take super long if pandas adoption isn't extremely tricky for some reason.  But we may have to spend a few weeks of work just to see whether or not it is difficult.  Are you interested in attempting that?

jbrockmendel commented:
> That is the question for you :). We could start on new DTypes for datetimes/timedelta. Obviously, it would be targeted to do whatever pandas wants

The big question mark this raises for me is "to what extent can numpy with new Dtypes be dropped in to existing pandas code and Just Work?"  Have you tried running the pandas test suite with the new Dtype system in place?

I'll be happy to serve as a guinea pig for the new Dtypes, but can't make any promises about them actually getting used inside pandas anytime soon.

> Figure out the basic requirements (timezones? resolution?). One way to solve this for me would be if we had "scalars" for blueprints (that could be pandas scalars or even datetime.datetime)
> Create a corresponding DType (this I can do, maybe not full featured, but with the basics there)

Setting aside my ignorance of what a attributes/methods a Dtype needs, the idea I've been pursuing has been to create a Localizer class that would look a little bit like a dtype if you squint.  It would be pinned as  private attribute on a Timestamp and on a DatetimeTZDtype, and its attributes would include a tzinfo and a NPY_DATETIMEUNIT.

My last attempt to implement Localizer as a class was https://github.com/pandas-dev/pandas/pull/46246/files#diff-23ab6ca878fcae5ed2cb46d90ad29a18c2c4f5ba18bf16918b91b8ef613afefdR537, and its utc_val_to_local_val method was the guts of it.  The problem I faced was that making that a method instead of repeated inline _really_ hurt performance (see asvs posted in the OP to that PR).

----

In the interim, is pushing forward with this PR worthwhile?

mattip commented:
>  [H]ow it compares to most of the API which uses the API table rather than relying on linking?

The API table is for NumPy C-API functions (like `PyArray_NDIM`), not for pure C functions (like the ones in mathlib). The random distribution functions are more similar to the mathlib ones. I guess there could be room for discussion around exporting these functions: are they NumPy C-API or C functions?

seberg commented:
Well, I am OK with this if Matti is.  `npymath` is its own library though and doesn't even need Python to be loaded?  This code does create Python errors and work with python `datetime`.  The header would also have to be in `core/src/include/numpy/*`, I think.

I have a bit of a bad feeling about linking against `multiarray`, though.  I am sure this cython example works on linux, but I don't know if we can link multiarray easily?  If that is indeed OK, then this is OK, but right now I don't quite trust that approach.

The alternative is to rename the functions `PyArray_...` to align them with most of our C-API, and put them in the C-API table like everything else.  They all have error returns I think, so we can reasonably deprecate them when we want to.
I somewhat expect that would be the way to go.  I don't have an opinion on whether this will be useful, so I am willing to just say that it should be more use than nuisance.

jbrockmendel commented:
> are they NumPy C-API or C functions?

I'm not sure what the distinction is.  The functions touched in this PR are all ones that pandas re-implements near-verbatim.

> The alternative is to rename the functions PyArray_... to align them with most of our C-API, and put them in the C-API table like everything else

If there's a non-trivial chance of some of these getting split out into a third package, maybe the future-proof thing to do would be to not put them into the "official" C-API?

seberg commented:
Lets mark it for discussion at next weeks meeting.

The distinction is about how we link the functions.  This PR links them like a normal dynamic library.  Which we do for some math and random functions.
But the rest of the NumPy C-API uses an API "table".  I.e. when you do `import_array()` there are some python (not C!) function calls that fill in a table of function pointers for you.  
The API functions like `PyArray_NewFromDescr(...)` are then actually macros expanding into something like `multiarray_api_table[9](...)`  (if this was function number 9).
So from within the NumPy Python module, there is currently no C linking at all.  Because of that, I am unsure that it is something we can do reliably.

jbrockmendel commented:
Noting as I go (in implementing non-nanosecond support in pandas) other numpy functions that I'd like to re-use rather than re-implement (will update this list if/when I find more):

get_datetime_metadata_from_dtype
npy_timedeltastruct

cc @WillAyd if im missing anything

WillAyd commented:
I think there might be a few in the JSON space as well we use for serialization:

https://github.com/pandas-dev/pandas/blob/main/pandas/_libs/src/ujson/python/date_conversions.h

jbrockmendel commented:
> Lets mark it for discussion at next weeks meeting.

which meeting(s) should i plan to attend?  this will be my first one.

seberg commented:
@jbrockmendel wednesday at 16UTC: https://hackmd.io/68i_JvOYQfy9ERiHgXMPvg.  I wanted to see if we can nail down the linking thing with Matti and others.  But happy if you come.  I am seriously considering making a (minimal) "new" Datetime dtype so that you can see what is missing for proper support in pandas.  My `unitdtype` prototype doesn't quite work, but maybe a datetime/timedelta DType will be an incitment to try that.

jbrockmendel commented:
zoom wouldn't let me into the meeting, which was probably OK because I was barely awake

seberg commented:
@jbrockmendel we discussed this briefly.  I think there is a tendency of "can't we remove this eventually?", but that doesn't matter too much...
The point is, if we are to expose this, I think we need to do it to the C-API table, which means:
* removing it from all headers (probably, not sure it is strictly necessary)
* Adding it to: https://github.com/numpy/numpy/blob/aaae9d1da5174290152eb380bf22cfd991089bda/numpy/core/code_generators/numpy_api.py
* Marking it with a comment starting with `/*NUMPY_API`
* I would want to rename the functions to `PyArray_*` or some similar pattern.
* All functions exposed like this should be able to return an error state, so that we can deprecate them easily if it comes to that.  (Exceptions may be possible on an individual basis, but right now it doesn't seem like it should be necessary.)

The rest should happen automatically, but you should copy paste that magic comment from somewhere, and if you get strange complication errors formatting can be the reason (e.g. a marked function has to have the initial `{` bracket on the next line I think).

As for a user DType, we did not really discuss it.  But I am happy to create a prototype for an external DType, so that you could see what modifications in pandas may make it work.  Something that would be useful in any case, probably.

jbrockmendel commented:
> I think there is a tendency of "can't we remove this eventually?", but that doesn't matter too much...

One option would be to move the duplicated code to a third package upstream of both numpy and pandas.  That's a complication I'd prefer to avoid for now, but if it makes things easier on the numpy devs, can revisit it.

> As for a user DType, we did not really discuss it. But I am happy to create a prototype for an external DType, so that you could see what modifications in pandas may make it work. Something that would be useful in any case, probably.

Based on our last phone call I got the impression that this is mostly-orthogonal to what I'm working on.  Which is great!  Because it means that I can spend time helping you with your project and feel less bad about taking up your time!

> [...] The rest should happen automatically, but you should copy paste that magic comment from somewhere [...]

OK, I think I get what you're asking.  In addition, I'm going to see if there's anything worth salvaging from #16364.

jbrockmendel committed:
salvage bis of #16364

seberg commented:
Its a bit annoying, basically, you list the function here:
https://github.com/numpy/numpy/blob/aaae9d1da5174290152eb380bf22cfd991089bda/numpy/core/code_generators/numpy_api.py#L356

The next will be `(307,)` because there is a 306 at the top (this is really just an enumeration).  The function itself has to be `NPY_NO_EXPORT ...`, but must have a comment right before the definition that is formatted as:
```
/*NUMPY_API
 *
 * Actual comment
 */
NPY_NO_EXPORT int
myfunction(...)
{
```
In principle this should only mean adding that `NUMPY_API` thing on the first line of the comment, assuming that there is already a code comment.  (Look e.g. at `PyArray_All` or any other C-API function.)

I think there may be a hash hardcoded somewhere that needs updating (i.e. compilation will complain), but I would have to `grep` for it also right now.

The header for a function marked with `/*NUMPY_API` is auto-generated, so you have to remove it from the current headers.

jbrockmendel commented:
Thanks for looking into this.  I've spread myself too thin and need to put this on the back-burner.  It _may_ also turn out that we will want to get access to parse_iso_8601_datetime.  Will revisit in a few weeks or months.

jbrockmendel commented:
Revisiting this, I think I got the numpy_api.py file thing sorted out.  Having trouble with the `__init__.pxd` files when using 

```
cdef extern from "numpy/datetime_strings.h":
# or
cdef extern from "numpy/mutltiarray/datetime_strings.h":
# or
cdef extern from "numpy/core/src/mutltiarray/datetime_strings.h":
```

Either way it says the file is not found.  The actual location of the file is numpy/core/src/multiarray/datetime_strings.h

bashtage commented:
Are your .h files located at numpy/core/include/numpy/<file-name>.h?  The `numpy/header.h` includes works since the include path is `numpy/core/include`

seberg commented:
@jbrockmendel you cannot include `numpy/mutltiarray/datetime_strings.h`, those are private header files.  Once added to the API (and recieving the magic comment), you should not need to add any additional includes.  See this comment: https://github.com/numpy/numpy/pull/21199#issuecomment-1076804447

jbrockmendel commented:
> you should not need to add any additional includes

I'm confused.  I need to add these to the `__init__.pxd` files somewhere, right

seberg commented:
@jbrockmendel no, if you have the special comment:
```
/*NUMPY_API
 *...
 */
```
you do not need to any new include.

jbrockmendel committed:
add NUMPY_API to comments, revert includes

jbrockmendel commented:
yah the tests are breaking because the functions aren't declared in `__init__.pxd`

> checks.pyx:45:7: cimported module has no attribute 'convert_datetime_to_datetimestruct'

seberg commented on numpy/__init__.cython-30.pxd:
```diff
@@ -806,6 +806,10 @@ cdef extern from "numpy/ndarraytypes.h":
 
+    ctypedef struct npy_datetimestruct:
+        int64_t year
+        int32_t month, day, hour, min, sec, us, ps, as
```
Oh, you do need to duplicate the function signatures here of course, but with the magic comment it is now part of this header (I think).

jbrockmendel replied:
Everything duplicated here is inside a `cdef extern from "numpy/foo.h":` block for some "foo".  what im trying to ask is what "foo" is here

seberg replied:
Foo is this block, I think.  Whatever all of the other functions are like `PyArray_FromArray`, etc.

jbrockmendel committed:
try suggestion

jbrockmendel committed:
fix annotation

jbrockmendel commented:
rebased, still no luck figuring out why the things added in `__init__.pxd` (and numpy_api.py) are showing up as non-existent at runtime.

WillAyd commented:
@jbrockmendel what do you mean by non-existent at runtime? Have an error to share?

jbrockmendel committed:
Merge branch 'main' into npy_yes_export

jbrockmendel commented:
Back in july I was seeing `checks.pyx:45:7: cimported module has no attribute 'convert_datetime_to_datetimestruct'`, but looking at the test logs now im just seeing segfaults.  Re-compiling locally is failing with `fatal error: 'x86-simd-sort/src/avx512-32bit-qsort.hpp' file not found #include "x86-simd-sort/src/avx512-32bit-qsort.hpp"` so i need to figure that out before i can try to replicate.

jbrockmendel commented:
Locally I'm seeing e.g.

```

    def test_get_datetime_iso_8601_strlen(install_temp):
        # GH#21199
>       import checks
E       ModuleNotFoundError: No module named 'checks'
```

with further-up traceback

```
    dts.day = dt.day
    dts.hour = dt.hour
    dts.sec = dt.second
    dts.us = dt.microsecond

    cnp.make_iso_8601_datetime(
      ^
------------------------------------------------------------

checks.pyx:65:7: cimported module has no attribute 'make_iso_8601_datetime'
```

with similar messages for convert_datetime_to_datetimestruct, get_datetime_iso_8601_strlen, npy_datetimestruct

WillAyd commented:
It looks like you are missing the `self` parameter in your function, so my guess is your parametrization is getting mangled

jbrockmendel commented:
> It looks like you are missing the self parameter in your function, so my guess is your parametrization is getting mangled

Good catch but no dice.

seberg commented on numpy/core/tests/examples/cython/checks.pyx:
```diff
@@ -30,3 +30,46 @@ def get_dt64_unit(obj):
+def convert_datetime_to_datetimestruct():
+    cdef:
+        cnp.npy_datetimestruct dts
+        cnp.PyArray_DatetimeMetaData *meta = NULL
```
I don't think you can pass NULL.

If it doesn't work locally, it sounds a bit like you are just picking up the old `__init__.pyd` while compiling `checks.pyx`?  Maybe try using `python runtests.py` just to see if it works?

jbrockmendel replied:
removed the NULL.

using python runtests.py im now getting segfaults locally.  that's progress sorta.

jbrockmendel replied:
that helped!  im also seeing compiler warnings about make_iso_8601_datetime (in checks.pyx)  having result and outlen unitialized.

jbrockmendel committed:
Fixup missing self

jbrockmendel committed:
fix tests

jbrockmendel committed:
lint fixup

jbrockmendel commented:
Tests passing locally, booyah!

@WillAyd will be proud of me, one of the steps was "googling how 'switch' statements work"

jbrockmendel committed:
Merge branch 'main' into npy_yes_export

charris commented on numpy/core/code_generators/cversions.txt:
```diff
@@ -68,4 +68,4 @@
 # Version 16 (NumPy 1.24) No change.
 # Version 16 (NumPy 1.25) No change.
-0x00000010 = 04a7bf1e65350926a0e528798da263c0
+0x00000010 = ad404c9986484ac22c9e9cd985f2eb1e
```
You should start a new block for 1.25, make it `0x00000011` (Version 17), note the reason, and keep the old hash for 1.23-1.24.

jbrockmendel replied:
looks like changing this broke the world.  do i need to change somehwere else to expect "17"?

charris replied:
Look at `doc/HOWTO_RELEASE.rst`, at line 187.

charris commented on numpy/core/code_generators/numpy_api.py:
```diff
@@ -370,6 +370,11 @@ def get_StealRef():
+    'convert_datetimestruct_to_datetime':   (308,),
+    'convert_pydatetime_to_datetimestruct': (309,),
+    'get_datetime_iso_8601_strlen':         (310,),
+    'make_iso_8601_datetime':               (311,),
```
These names are inconsistent with the previous naming convention, camel case and Py prefix. Probably should put some work into the name , Maybe start with `PyDateTime_:)

jbrockmendel replied:
not my call, but ill note that as the intended consumer the python-style names are nice.  my nitpick is id like to change "datetime" -> "datetime64" in the first two convert_foo functions here.

seberg replied:
Maybe `NpyDateTime` is best, we do have `NpyIter` as priot art or just `NpyTime_` even.  Even just `npy_` might be OK, although that is really only used for internal functions as a pattern.
Others do use `PyArray_` like `PyArray_DatetimeToDatetimeStruct`.

`PyDateTime_` is owned used by the builtin datetime module, so we can't use it.

I am not in the mood for bike-shedding the prefix or camel vs. non-camel case, but we need a prefix in C.

Maybe there is a typical pattern to create thin wrapper functions in cython to get shorter names, since I can understand the dislike for C-style names in cython cimports.

jbrockmendel replied:
So...

NpyDatetime_ConvertDatetime64ToDatetimeStruct
NpyDatetime_ConvertPydatetimeToDatetimeStruct
NpyDatetime_GetDatetimeISO8601StrLen
NpyDatetime_MakeISO8601Datetime

?

seberg replied:
I would capitalize the T, but that would work for me.  @charris do you have any preference?

jbrockmendel replied:
updated+green, though some whitespace is less pretty than it used to be.  not sure if anyone cares about that

jbrockmendel replied:
is the whitespace a blocker here?

charris replied:
The file is hashed for cversion, so it is best that it looks nice enough that folks aren't tempted to "fix" it. Apart from the ragged alignment, I wonder if the comment is needed?

charris commented on numpy/core/src/multiarray/datetime.c:
```diff
@@ -3910,8 +3913,7 @@ datetime_to_timedelta_resolve_descriptors(
         PyArray_DTypeMeta *dtypes[2],
         PyArray_Descr *given_descrs[2],
-        PyArray_Descr *loop_descrs[2],
-        npy_intp *NPY_UNUSED(view_offset))
```
This looks suspicious.

jbrockmendel replied:
this looks unrelated, no idea what happened here.  will revert.

jbrockmendel committed:
revert accident, update cversions.txt

jbrockmendel committed:
update C_API_VERSION

jbrockmendel committed:
lint fixup

jbrockmendel commented:
green, booyah!

jbrockmendel committed:
add NpyDatetime_ to names

jbrockmendel committed:
whitespace fixup

seberg committed:
MAINT: Modify style, change spelling a bit and fix Cython except

seberg commented on numpy/core/src/multiarray/_datetime.h:
```diff
@@ -286,27 +268,6 @@ convert_datetime_to_pyobject(npy_datetime dt, PyArray_DatetimeMetaData *meta);
-NPY_NO_EXPORT int
-convert_datetimestruct_to_datetime(PyArray_DatetimeMetaData *meta,
-                                    const npy_datetimestruct *dts,
-                                    npy_datetime *out);
```
If we move things to the public API, we don't really need the internal def's, we don't duplicate it for others as well.  So just fully removed them here.

seberg commented on numpy/__init__.cython-30.pxd:
```diff
@@ -855,6 +861,31 @@ cdef extern from "numpy/arrayscalars.h":
+            npy_datetimestruct *out) except -1
+    int convert_datetimestruct_to_datetime64 "NpyDatetime_ConvertDatetimeStructToDatetime64"(
+            PyArray_DatetimeMetaData *meta, const npy_datetimestruct *dts,
+            npy_datetime *out) except -1
```
The `except -1` was missing (and noexcept in the one case to be explicit).  So added those (and prefer this formatting, but don't want to bike-shed it).

seberg committed:
MAINT: Remove noexcept again

It is not supported on some older cython versions.  But even for
Cython 3 it should not matter in an extern C blocks...

seberg commented:
@charris could you have a quick pass if you prefer other name spellings?  I would agree they are terribly unwieldy and impossible to remember with camel case or not.  But I guess at least they will be rarely used...
Happy to change though.  Otherwise, as much as I dislike adding API, these seem OK to expose and I suspect the change generally fine.

I think this is otherwise ready, although as an API addition should get a changelog entry, and maybe stubs in the actual documentation or copy-pasting the doc string.  (that should be automated, but I don't think it is quite...)

jbrockmendel commented:
Thanks for pushing @seberg!  I'm excited to get this in.

If you have a dislike of adding API and i have a couple more things I'd like to upstream would you suggest a) trying to include those here, b) separate PR, or c) be happy with this and don't push my luck?

In particular I'm thinking of https://github.com/pandas-dev/pandas/blob/main/pandas/_libs/tslibs/src/datetime/np_datetime.c#L1068 which I have not been able to imlpement directly in cython.

seberg commented:
Lets follow up with others, after the pattern is there it should be easier.  The main point is, I only want to add functions that are clearly stable and not weird to support.  These all seem OK (we need them, no obvious API Flux seems likely).

In this case, the weirdest part of the API is the timezone handling in the printing.  I don't think we use it in NumPy, do you require it for pandas, because otherwise we should remove it.

jbrockmendel commented:
>  the weirdest part of the API is the timezone handling in the printing. I don't think we use it in NumPy, do you require it for pandas, because otherwise we should remove it.

Our originally-vendored version of make_iso_8601_datetime removes the local, tzoffset, and casting keywords, still has "utc".

Looks like in convert_pydatetime_to_datetimestruct the numpy version has a 2016-era `if (DEPRECATE(` about parsing timezones which ours does not.  If you were to enforce that deprecation, we'd need to re-implement or otherwise find a workaround.  Based on the two places where we use it, it tentatively looks like a workaround wouldn't be the end of the world.

Are there are other places you have in mind?

seberg commented:
Marking for 1.25.x.  This will probably need a rebase before the end (C version update).  The API also will have to be marked with the `MinVer("1.25")` after we change the C-API backcompat strategy (meaning these would just fail to compile in Cython without the correct macro).

However, if anyone wants to review the C-API addition itself, it would be good to do so.  Otherwise I will just put this in at some point before branching 1.25 probably, since it sounds like pandas needs it for better or worse.

charris commented:
@jbrockmendel  Needs a rebase.

seberg committed:
Merge branch 'main' into npy_yes_export

seberg committed:
MAINT: Add MinVersion to API additions

seberg commented on numpy/core/code_generators/numpy_api.py:
```diff
@@ -370,6 +370,11 @@ def get_annotations():
+    'NpyDatetime_ConvertDatetimeStructToDatetime64': (308, MinVersion("1.25")),
+    'NpyDatetime_ConvertPyDateTimeToDatetimeStruct': (309, MinVersion("1.25")),
+    'NpyDatetime_GetDatetimeISO8601StrLen':          (310, MinVersion("1.25")),
+    'NpyDatetime_MakeISO8601Datetime':               (311, MinVersion("1.25")),
```
Added the min-version and hopefully resolved a bit of a merge/rebase mess (the sorting submodule was updated in the diff).

seberg committed:
TST: Make sure cython tests define NPY_TARGET_VERSION to >=1.25

seberg commented on numpy/core/tests/examples/cython/setup.py:
```diff
@@ -9,7 +9,11 @@
+macros = [
+    ("NPY_NO_DEPRECATED_API", 0),
+    # Require 1.25+ to test datetime additions
+    ("NPY_TARGET_VERSION", "NPY_1_25_API_VERSION"),
```
Have to add this, since NumPy now defaults to not exporting these new symbols otherwise (leading to compilation warnings and runtime error if you use the new functions).

mattip commented:
This looks good. It needs a release note.

charris commented:
Look in `doc/release/upcoming_changes` to see how the release notes are done.

charris commented:
These functions also need documentation, probably add a page in `doc/source/reference/c-api`. Given that I would like to branch today or tomorrow, I'm going to push this off to the next release, which will require some additional changes.

charris commented:
> but I had assumed we can push this through.

I could compromise on a release note.

jbrockmendel commented:
I tried rebasing this morning, had a merge conflict, and got discouraged.

> I don't know if @jbrockmendel cares, but I wouldn't mind pushing this through

Getting this in for 1.2x doesn't make much difference for me.  Getting it in by 2.0 is a bigger deal.

ngoldbaum commented:
I just tried writing a cast from the string dtype I've been working on to numpy datetimes and realized it would be nice to have `parse_iso_8601_datetime` exposed in the C API. @jbrockmendel would you mind if I pushed some commits to this PR to expose it and add it to the cython bindings? Also happy to fix the other issues raised by reviewers so this can go in sooner rather than later.

jbrockmendel commented:
That sounds great.

ngoldbaum committed:
Merge branch 'main' into npy_yes_export

ngoldbaum committed:
MAINT: remove one more unnecessary header prototype

ngoldbaum committed:
DOC: add release note

ngoldbaum committed:
API: Add NpyDatetime_ParseISO8601Datetime and update target version

ngoldbaum committed:
DOC: add docs for the datetime C API

ngoldbaum commented:
Merged with main, added release notes, and new C API datetime docs. If there are any test failures I'll clean those up.

Is there a way to see the doc build generated by the CI job for a PR?

ngoldbaum commented on numpy/core/code_generators/cversions.txt:
```diff
@@ -73,4 +73,6 @@
 # Version 18 (NumPy 2.0.0)
-0x00000012 = 5af92e858ce8e95409ae1fcc8f508ddd
+# Many API deprecations have been finalized
+# Add datetime conversion functions GH#21199
```
If we're not enumerating every single change for numpy 2.0 I'm happy to remove these.

seberg replied:
Yah, lets not, already a few removals too.

ngoldbaum committed:
MAINT: fix name of release note

ngoldbaum commented:
Oops, forgot the cython binding for `parse_iso_8601_datetime`, will add that too.

ngoldbaum committed:
MAINT: remove comments from cversions.txt for numpy 2.0

ganesh-k13 commented:
> Is there a way to see the doc build generated by the CI job for a PR?

https://output.circle-artifacts.com/output/job/2a322bc7-156b-4c32-b269-ca21c17a5534/artifacts/0/doc/build/html/index.html

Click on `details` of `ci/circleci: build artifacts`

seberg commented:
Time to put this in, docs looked good without checking every tiny detail, thanks @jbrockmendel and @ngoldbaum for pushing it over the finish line!

jbrockmendel commented:
nice!  thanks reviewers for walking me through this and @ngoldbaum for getting it across the finish line.  Looking forward to ripping a bunch of this out of pandas next year!

jbrockmendel commented:
> I still would prefer we declare datetime64 to be a user-defined dtype, and split the code out into a separate PyPI-installable package. This would be a good test of the new dtype machinery as well as clean up NumPy internals.

I'd be interested in particular if we could get pyarrow to share the implementation, since install size has been coming up as a pain point recently.  @mattip are you the person to talk to about making this a reality?

mattip commented:
I would think that in parallel to the [string dtype](https://github.com/numpy/numpy-user-dtypes/tree/main/stringdtype) that is progressing quite nicely, a datetime64-like dtype could take form there. It is something to discuss at a community meeting: who are the interested parties to design the underlying structures and how they find the time (haha) to make progress.
